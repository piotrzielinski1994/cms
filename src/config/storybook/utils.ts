import { ReactNode } from 'react';

const DEFAULT_VALUE = '__default__' as const;
const THUMBNAIL_ID = 'thumbnail';

const storyToUrlPath = (title: string) => {
  const parts = title.split('/');
  const story = parts[parts.length - 1];
  const titlePart = title.toLowerCase().replace(/\//g, '-').replace(/\s+/g, '-');
  const storyPart = story
    .replace(/([a-z0-9])([A-Z])/g, '$1-$2')
    .replace(/([a-zA-Z])(\d+)/g, '$1-$2')
    .toLowerCase();
  return `${titlePart}--${storyPart}`;
};

const getFallback = <T extends ReactNode>(value: T, translation: string): T => {
  return value === DEFAULT_VALUE ? (translation as T) : value;
};

export { DEFAULT_VALUE, getFallback, storyToUrlPath, THUMBNAIL_ID };
