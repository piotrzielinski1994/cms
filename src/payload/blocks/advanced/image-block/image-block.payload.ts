import { AdminTranslations } from '@/config/store/locales.config';
import { Block } from 'payload';
import thumbnail from './image-block.webp';

const imageBlock1BlockPayloadConfig = {
  slug: 'image-block-1',
  interfaceName: 'ImageBlockBlock',
  imageURL: thumbnail.src,
  fields: [
    {
      name: 'image',
      type: 'group',
      label: ({ t }: { t: AdminTranslations }) => t('fields:image'),
      fields: [
        {
          type: 'row',
          fields: [
            {
              name: 'default',
              type: 'upload',
              relationTo: 'images',
              required: true,
              label: ({ t }: { t: AdminTranslations }) => t('common:default'),
              admin: {
                width: '50%',
              },
            },
            {
              name: 'dark',
              type: 'upload',
              relationTo: 'images',
              label: ({ t }: { t: AdminTranslations }) => t('common:dark'),
              admin: {
                width: '50%',
              },
            },
          ],
        },
      ],
    },
    {
      name: 'isReversed',
      type: 'checkbox',
      label: ({ t }: { t: AdminTranslations }) => t('fields:isReversed'),
    },
    {
      name: 'heading',
      type: 'text',
      localized: true,
      label: ({ t }: { t: AdminTranslations }) => t('fields:heading'),
    },
    {
      name: 'subheading',
      type: 'text',
      required: true,
      localized: true,
      label: ({ t }: { t: AdminTranslations }) => t('fields:subheading'),
    },
    {
      name: 'buttons',
      type: 'array',
      label: ({ t }: { t: AdminTranslations }) => t('fields:buttons'),
      admin: {
        initCollapsed: true,
      },
      fields: [
        {
          name: 'label',
          type: 'text',
          required: true,
          localized: true,
          label: ({ t }: { t: AdminTranslations }) => t('fields:label'),
        },
        {
          name: 'reference',
          type: 'relationship',
          label: ({ t }: { t: AdminTranslations }) => t('fields:documentToLinkTo'),
          relationTo: ['pages'],
          required: true,
        },
        {
          name: 'selector',
          type: 'text',
          label: ({ t }: { t: AdminTranslations }) => t('fields:selector'),
        },
      ],
    },
  ],
} satisfies Block;

export { imageBlock1BlockPayloadConfig };
