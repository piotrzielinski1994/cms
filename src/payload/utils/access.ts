import { User } from '@/payload.types';
import { Access, FieldAccess } from 'payload';

const anyone: Access = () => true;

const authenticated: Access = ({ req: { user } }) => {
  return Boolean(user);
};

const admin: Access = ({ req: { user } }) => {
  return Boolean(user?.roles?.includes('admin'));
};

const adminOrSelf: Access = ({ req: { user } }) => {
  if (!user) return false;
  if (user.roles?.includes('admin')) return true;
  return { id: { equals: user.id } };
};

const adminFieldLevel: FieldAccess<{ id: string }, User> = ({ req: { user } }) => {
  return Boolean(user?.roles?.includes('admin'));
};

const authenticatedOrPublished: Access = ({ req: { user } }) => {
  if (user) return true;
  return { _status: { equals: 'published' } };
};

export { admin, adminFieldLevel, adminOrSelf, anyone, authenticated, authenticatedOrPublished };
