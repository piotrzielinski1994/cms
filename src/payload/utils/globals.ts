import type { Config } from '@/payload.types';
import configPromise from '@/payload/payload.config';
import { Locale } from 'next-intl';
import { unstable_cache } from 'next/cache';
import { getPayload } from 'payload';

const getGlobal = async (slug: keyof Config['globals'], locale: Locale = 'en', depth = 1) => {
  const payload = await getPayload({ config: configPromise });
  return payload.findGlobal({ slug, depth, locale });
};

const getCachedGlobal = (slug: keyof Config['globals'], locale: Locale, depth = 1) => {
  const cacheConfig = { tags: [`global__${locale}_${slug}`] };
  return unstable_cache(() => getGlobal(slug, locale, depth), [slug, locale], cacheConfig);
};

export { getCachedGlobal };
