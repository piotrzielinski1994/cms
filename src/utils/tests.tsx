import { Providers } from '@/providers';
import { ComponentProps, ReactNode } from 'react';

const resolveServerComponent = async <T extends (props: unknown) => Promise<ReactNode>>(
  Component: T,
  props: Parameters<T>[0],
) => {
  const ComponentResolved = await Component(props);
  return () => ComponentResolved;
};

const withProviders = (
  children: ReactNode,
  props: Partial<Omit<ComponentProps<typeof Providers>, 'children'>> = {},
) => {
  const overwrittenProps: ComponentProps<typeof Providers> = {
    locale: 'en',
    initialTheme: 'light',
    initialFontScale: 'base',
    initialCookiesConsent: true,
    ...props,
    children,
  };
  return <Providers {...overwrittenProps} />;
};

export { resolveServerComponent, withProviders };
