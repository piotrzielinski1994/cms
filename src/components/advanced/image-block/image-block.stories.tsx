import { getThemeConfig } from '@/config/store/themes.config';
import { StoryContext } from '@/config/storybook/components';
import {
  DEFAULT_VALUE,
  getFallback,
  imagesPerColorPref,
  THUMBNAIL_ID,
} from '@/config/storybook/utils';
import type { Meta, StoryObj } from '@storybook/react';
import { useTranslations } from 'next-intl';
import { type ComponentProps } from 'react';
import { ImageBlock as ImageBlockComponent } from './image-block';

type Args = ComponentProps<typeof ImageBlockComponent>;

const meta = {
  component: ImageBlockComponent,
  title: 'Advanced/ImageBlock',
  argTypes: {
    isReversed: { control: 'boolean' },
    image: { control: 'object' },
    heading: { control: 'text' },
    subheading: { control: 'text' },
    buttons: { control: 'object' },
  },
  args: {
    isReversed: false,
    image: {
      src: DEFAULT_VALUE,
      alt: DEFAULT_VALUE,
      width: undefined,
      height: undefined,
    },
    heading: DEFAULT_VALUE,
    subheading: DEFAULT_VALUE,
    buttons: [
      { label: DEFAULT_VALUE, href: '/path-1', variant: 'primary' },
      { label: DEFAULT_VALUE, href: '/path-2', variant: 'secondary' },
    ],
  },
} satisfies Meta<Args>;

const Render = ({ image, heading, subheading, buttons, ...args }: Args, context) => {
  const { theme } = context.globals as StoryContext['globals'];
  const t2 = useTranslations('fields');
  const tButton = useTranslations('storybook.basic.button');
  const themedImage = imagesPerColorPref[getThemeConfig(theme).colorPreference];

  return (
    <ImageBlockComponent
      {...args}
      id={THUMBNAIL_ID}
      image={{
        ...image,
        src: getFallback(image.src, themedImage.src),
        alt: getFallback(image.alt, t2('image')),
      }}
      heading={getFallback(heading, t2('heading'))}
      subheading={getFallback(subheading, t2('subheading'))}
      buttons={buttons?.map((it, index) => ({
        ...it,
        label: getFallback(it.label, `${tButton('default')} ${index + 1}`),
      }))}
    />
  );
};

const ImageBlock: StoryObj<typeof ImageBlockComponent> = { render: Render };

export { ImageBlock };
export default meta;
