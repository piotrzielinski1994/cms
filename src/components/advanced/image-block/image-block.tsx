'use client';

import { ButtonLink } from '@/components/basic/button/button';
import { Image as BasicImage } from '@/components/basic/image/image';
import { cn } from '@/utils/tailwind';
import { StaticImageData } from 'next/image';
import { ComponentProps, HTMLAttributes } from 'react';

type ImageBlockProps = HTMLAttributes<HTMLDivElement> & {
  isReversed?: boolean;
  image: Omit<StaticImageData, 'width' | 'height'> & {
    alt: string;
    width?: number;
    height?: number;
  };
  heading?: string;
  subheading?: string;
  buttons?: Array<Pick<ComponentProps<typeof ButtonLink>, 'href' | 'variant'> & { label: string }>;
};

const ImageBlock = ({
  isReversed,
  image,
  heading,
  subheading,
  buttons = [],
  ...props
}: ImageBlockProps) => {
  return (
    <div {...props} className="grid md:grid-cols-2 gap-4 md:gap-8">
      <div className={cn('order-1 md:order-2', { 'md:order-1': isReversed })}>
        <BasicImage
          className="bg-background1 min-h-full"
          src={image.src}
          alt={image.alt}
          aspectRatio={
            Boolean(image.width && image.height)
              ? { width: image.width!, height: image.height! }
              : undefined
          }
        />
      </div>
      <div
        className={cn('grid justify-items-start content-center gap-4', 'order-2 md:order-1', {
          'md:order-2': isReversed,
        })}
      >
        <h3 className="text-4xl font-semibold">{heading}</h3>
        {subheading && <p>{subheading}</p>}
        {buttons.length > 0 && (
          <div className="flex gap-4">
            {buttons.map((button) => {
              return (
                <ButtonLink key={button.label} href={button.href} variant={button.variant}>
                  {button.label}
                </ButtonLink>
              );
            })}
          </div>
        )}
      </div>
    </div>
  );
};

export { ImageBlock };
