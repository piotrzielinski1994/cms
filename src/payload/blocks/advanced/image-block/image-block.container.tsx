'use client';

import { ImageBlock } from '@/components/advanced/image-block/image-block';
import { getThemeConfig } from '@/config/themes.config';
import { ImageBlockBlock, Image as ImageModel, Page } from '@/payload.types';
import { useThemeStore } from '@/store/theme';

const ImageBlock1Container = (props: ImageBlockBlock) => {
  const { isReversed, image, heading, subheading, buttons } = props;
  const theme = useThemeStore((store) => store.theme);
  const { default: defaultImage, dark: darkImage } = image as {
    default: ImageModel;
    dark?: ImageModel;
  };
  const prefersDark = getThemeConfig(theme).colorPreference === 'dark';
  const imageToShow = !prefersDark ? defaultImage : (darkImage ?? defaultImage);

  return (
    <ImageBlock
      isReversed={isReversed ?? undefined}
      image={{
        src: imageToShow.url ?? '',
        alt: imageToShow.alt,
        width: imageToShow.width ?? undefined,
        height: imageToShow.height ?? undefined,
      }}
      heading={heading ?? undefined}
      subheading={subheading ?? undefined}
      buttons={buttons?.map((button, index) => {
        const href = (button?.reference?.value as Page).path ?? '';
        const variant = index === 0 ? 'primary' : 'secondary';
        return { label: button.label, href, variant };
      })}
    />
  );
};

export { ImageBlock1Container };
