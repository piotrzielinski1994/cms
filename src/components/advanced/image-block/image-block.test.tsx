import { withProviders } from '@/utils/tests';
import { render, waitFor } from '@testing-library/react';
import { ComponentProps } from 'react';
import { describe, expect, it } from 'vitest';
import { axe } from 'vitest-axe';
import { ImageBlock } from './image-block';

describe('ImageBlock', () => {
  const defaultProps = {
    heading: 'Heading',
    subheading: 'Subheading',
    buttons: [
      {
        label: 'Primary Button',
        href: '/primary',
      },
      {
        variant: 'secondary',
        label: 'Secondary Button',
        href: '/secondary',
      },
    ],
    image: {
      src: '/images/placeholder.webp',
      alt: 'Image',
      width: 1920,
      height: 1080,
    },
  } satisfies ComponentProps<typeof ImageBlock>;

  it('should have no accessibility violations', async () => {
    const { container } = await waitFor(() =>
      render(withProviders(<ImageBlock {...defaultProps} />)),
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
  return;
  it('should match the snapshot', async () => {
    const { container } = render(withProviders(<ImageBlock {...defaultProps} />));
    expect(container).toMatchSnapshot();
  });
});
