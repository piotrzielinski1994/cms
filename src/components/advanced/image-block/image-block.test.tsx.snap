// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ImageBlock > should match the snapshot 1`] = `
<div>
  <div
    class="grid md:grid-cols-2 gap-4 md:gap-8"
  >
    <div
      class="order-1 md:order-2"
    >
      <img
        alt="Image"
        class="object-cover cms-image bg-background1 min-h-full"
        data-nimg="1"
        decoding="async"
        height="1080"
        loading="lazy"
        sizes="100vw"
        src="/_next/image?url=%2Fimages%2Fplaceholder.webp&w=3840&q=75"
        srcset="/_next/image?url=%2Fimages%2Fplaceholder.webp&w=640&q=75 640w, /_next/image?url=%2Fimages%2Fplaceholder.webp&w=750&q=75 750w, /_next/image?url=%2Fimages%2Fplaceholder.webp&w=828&q=75 828w, /_next/image?url=%2Fimages%2Fplaceholder.webp&w=1080&q=75 1080w, /_next/image?url=%2Fimages%2Fplaceholder.webp&w=1200&q=75 1200w, /_next/image?url=%2Fimages%2Fplaceholder.webp&w=1920&q=75 1920w, /_next/image?url=%2Fimages%2Fplaceholder.webp&w=2048&q=75 2048w, /_next/image?url=%2Fimages%2Fplaceholder.webp&w=3840&q=75 3840w"
        style="color: transparent;"
        width="1920"
      />
    </div>
    <div
      class="grid justify-items-start content-center gap-4 order-2 md:order-1"
    >
      <h3
        class="text-4xl font-semibold"
      >
        Heading
      </h3>
      <p>
        Subheading
      </p>
      <div
        class="flex gap-4"
      >
        <a
          class="px-6 py-3 border-2 border-primary hover:border-primary/90 font-semibold whitespace-nowrap disabled:cursor-not-allowed focus-visible:tw-cms-outline bg-primary hover:bg-primary/90 text-primary-foreground disabled:border-primary/0 disabled:bg-primary/50"
          href="/primary"
        >
          Primary Button
        </a>
        <a
          class="px-6 py-3 border-2 border-primary hover:border-primary/90 font-semibold whitespace-nowrap disabled:cursor-not-allowed focus-visible:tw-cms-outline text-primary disabled:border-primary/50 disabled:text-primary/50 bg-background [&:not(:disabled)]:hover:bg-foreground/5"
          href="/secondary"
        >
          Secondary Button
        </a>
      </div>
    </div>
  </div>
</div>
`;
